codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/ProtocolFunctions.kt

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.rust.codegen.core.smithy.protocols
+
+import software.amazon.smithy.model.shapes.DocumentShape
+import software.amazon.smithy.model.shapes.MemberShape
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.model.shapes.Shape
+import software.amazon.smithy.rust.codegen.core.rustlang.Attribute
+import software.amazon.smithy.rust.codegen.core.rustlang.EscapeFor
+import software.amazon.smithy.rust.codegen.core.rustlang.RustModule
+import software.amazon.smithy.rust.codegen.core.rustlang.RustReservedWords
+import software.amazon.smithy.rust.codegen.core.rustlang.RustWriter
+import software.amazon.smithy.rust.codegen.core.smithy.CodegenContext
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType
+import software.amazon.smithy.rust.codegen.core.smithy.RustSymbolProvider
+import software.amazon.smithy.rust.codegen.core.smithy.contextName
+import software.amazon.smithy.rust.codegen.core.util.toSnakeCase
+
+/**
+ * Similar to [Writable], but provides the function name that is being generated as an argument.
+ */
+typealias ProtocolFnWritable = RustWriter.(String) -> Unit
+
+/**
+ * Utilities for generating protocol serialization/deserialization functions, and common support functions.
+ *
+ * This class should be used for generating all inline functions from a protocol code generator, as it is
+ * responsible for correctly organizing and code splitting those functions into smaller modules.
+ */
+class ProtocolFunctions(
+    private val codegenContext: CodegenContext,
+) {
+    companion object {
+        private val serDeModule = RustModule.pubCrate("protocol_serde")
+
+        fun commonFn(fnName: String, block: ProtocolFnWritable): RuntimeType =
```

Comment:
`crossOperationFn`? it wasn't immediately obvious to me what "common" meant here